### PR TITLE
Add type equality checks and overrides

### PR DESCRIPTION
**Pull request by Turnerj**
All classes implement IEquatable and have GetHashCode and Equals overrides. Not sure if it is worth adding the operator overloads for equality too.

### CODE REVIEW

File: Source/Schema.NET.Tool/ViewModels/Class.cs [hunk 1]
```diff
@@ -150,43 +150,72 @@ public override string ToString()
 
             stringBuilder.AppendIndentLine(4, "{");
 
-            if (string.Equals(this.Name, "Thing", StringComparison.OrdinalIgnoreCase))
```
Comment (RehanSaeed) [inline Source/Schema.NET.Tool/ViewModels/Class.cs]: Any reason for removing this property?

File: Source/Schema.NET/JsonLdObject.cs [hunk 2]
```diff
@@ -43,6 +43,32 @@ public class JsonLdObject
...
+
+        /// <inheritdoc/>
+        public bool Equals(JsonLdObject other)
+        {
+            if (other is null)
+            {
+                return false;
+            }
+
+            if (ReferenceEquals(this, other))
```
Comment (RehanSaeed) [inline Source/Schema.NET/JsonLdObject.cs]: In theory StyleCop would require `this.` before `ReferenceEquals`.

File: Tests/Schema.NET.Test/ThingTest.cs [hunk 3]
```diff
@@ -28,5 +28,71 @@ public class ThingTest
             Assert.Equal(
                 this.thing.ToString(),
                 JsonConvert.DeserializeObject<Thing>(this.json, TestDefaults.DefaultJsonSerializerSettings).ToString());
+
```
Comment (RehanSaeed) [inline Tests/Schema.NET.Test/ThingTest.cs]: Some derived type and mixed comparison tests would also be useful.

File: Source/Schema.NET.Tool/ViewModels/Class.cs (hunk 1 above)
Comment (Turnerj) [reply Source/Schema.NET.Tool/ViewModels/Class.cs]: If it exists, it leads to an unbuildable project. `Context` already exists on `JsonLdObject` and is type `JsonLdContext`. The override declares the type as `string` (this is why it fails) and the schema URL. On top of that, `JsonLdContext` already defines the default value for the context with the HTTPS URL.

So even sorting the type issue, it seems a little redundant having `Thing` specifically set a value that is already defined by default.

File: Source/Schema.NET/JsonLdObject.cs (hunk 2 above)
Comment (Turnerj) [reply Source/Schema.NET/JsonLdObject.cs]: `ReferenceEquals` is a static method so I don't think we can actually use `this`. We could use `object` if we wanted though Visual Studio was telling me to remove that.

**Comment by RehanSaeed** [comment]
Cool thanks.

We should bump the version. If we get some of your other changes in, it'd be a major version upgrade.

**Comment by Turnerj** [comment]
Yeah - if we can my overhaul of the `ValueJsonConverter` merged (once we've sorted any other loose ends for it out), I think we'd be in a good place to version bump.

**Pull request merged**